specs/WebResourceResponseReceived.md

```diff
@@ -0,0 +1,232 @@
+# Background
+The WebView2 team has been asked for an API to get the response for a web
+resource as it was received and to provide request headers not available when
+`WebResourceRequested` event is raised (such as HTTP Authentication headers).
+The `WebResourceResponseReceived` event provides such response representation
+and exposes the request as committed.
+
+In this document we describe the new API. We'd appreciate your feedback.
+
+# Description
+The `WebResourceResponseReceived` event allows developers to inspect the
+response object from URL requests (such as HTTP/HTTPS, file and data). A key
+scenario is to allow developers to get Auth headers from an HTTP response to
+authenticate other tools they're using, since the Auth headers are not exposed
+in the `WebResourceRequested` event.
+
+This event is raised when the WebView receives the response for a request for a
+web resource. It provides access to both the response as it was received and the
+request as it was committed, including modifications made by the network stack
+(such as the adding of HTTP Authorization headers). The app can use this event
+to view the actual request and response for a web resource, but modifications
+made to these objects are ignored.
+
+When the event is raised, the WebView will pass a
+`WebResourceResponseReceivedEventArgs`, which lets the app view the request and
+response. The additional `PopulateResponseContent` API is exposed from the event
+arguments so the app can get the response's body (if it has one). If the app
+tries to get the response content before the first call to
+`PopulateResponseContent` completes, the stream object returned will be null.
```

Comment:
I can see somebody scratching their head for days trying to figure out why they aren't getting any stream data from the response. Should we throw an exception in this case? It is, after all, a programming error. `E_NOT_VALID_STATE`? ("The group or resource is not in the correct state to perform the requested operation.")

Reply:
Definitely confusing, and this mitigation would have a property getter throwing an exception, so better than the head scratching but still not good API. It seems clear that the response just isn't meant to be a property. It looks like we're trying to economize on types by not creating a version of CoreWebView2WebResourceResponse that doesn't have a sync Content property, but it would be better to just create a new type.

Reply:
We did try to reuse the familiar type we have that represents an HTTP response. The idea is content won't be used that often for this API and in our implementation we have to do a separate async request for us that can take a relatively long time to get the content. 

One potential solution we could have is since the Content property is a stream, could we immediately return a stream and block on its read until data is available? Optionally, anyone that needs async behavior could still call PopulateResponseContent to know when it's fully available before reading the stream.

If that's not an acceptable solution, we will create a new type.

Reply:
Cagri, Luis and, I talked about this. We have a way to remove PopulateResponseContent for WinRT without adding anything else, but not for our COM API, and not sure about the .NET API.

### Context
In the WebResourceRequested event, the Response property is used by the caller to provide a fake response without the request actually going to the wire. In this case the Content property on the Response is only really set (sync) by the caller.

In the WebResourceResponseReceived event, the body of the response isn't available when the event dispatches - the event dispatches when the HTTP response headers have been read but not yet the body. This event is not blocking and is informing the end developer of the response. The Content property here is only usefully a get (async) by the caller.

Using the same response type for both means a response obtained via WebResourceResponseReceived can be used later with WebResourceRequested.

### Option 1

We remove the event arg's PopulateResponseContent method, leave the response class Content property as is, and in COM only add a Response method `Task<Stream> TryGetContentAsync()`.

In COM the IStream interface has only sync methods so we'd need the TryGetContentAsync for our COM API. In WinRT the stream interface uses async methods to read and so we could have our own stream wrapper implementation that under the covers does the work of TryGetContentAsync to get the actual underlying stream and then just call through. If we do that we don't need the TryGetContentAsync method in WinRT. The three of us discussing this aren't .NET veterans to know if we can do similarly with .NET. We see the .NET stream has both async and sync read methods.

If we can remove PopulateResponseContent everywhere and have the Content property just work properly in WinRT (& maybe in .NET) but not in COM where you need a separate TryGetContentAsync, do we still want a separate response class? As mentioned above its somewhat useful to use the same type between the two events in order to reuse a response

### Option 2

Make a new CoreWebView2WebResourceReadOnlyResponse that is the same as CoreWebView2WebResourceResponse but none of the properties have setters, there is no Content property, and we add the `Task<Stream> TryGetContentAsync()` method. We use the new CoreWebView2WebResourceReadOnlyResponse only with WebResourceResponseReceived and the old CoreWebView2WebResourceResponse only with WebResourceRequested.

Reply:
Option1 seems to be going through a lot of complication to avoid creating the new type in Option 2. Is it that difficult to create a type?

Reply:
No its just the types represent the same thing conceptually and using the same type would allow for easily using the two events together so seems like it should work that way.

Reply:
I like Option 2; there's no special method required to make the Content property start working, and from the app point of view, there's no benefit to having APIs that don't work. (The exception to that latter part though is if it's common for this type to be given to to utility processing code that doesn't need to know which stage the request is in.)